artichoke-backend/src/extn/core/time/offset.rs

```diff
@@ -0,0 +1,211 @@
+//! Parser for Ruby Time offset parameter to help generate `Time`.
+
+use crate::convert::{implicitly_convert_to_int, implicitly_convert_to_string};
+use crate::extn::core::symbol::Symbol;
+use crate::extn::core::time::Offset;
+use crate::extn::prelude::*;
+
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_OFFSET: f64 = i32::MAX as f64;
+#[allow(clippy::cast_precision_loss)]
+const MIN_FLOAT_OFFSET: f64 = i32::MIN as f64;
+
+impl TryConvertMut<Value, Option<Offset>> for Artichoke {
+    type Error = Error;
+
+    fn try_convert_mut(&mut self, options: Value) -> Result<Option<Offset>, Self::Error> {
+        let hash: Vec<(Value, Value)> = self.try_convert_mut(options)?;
+
+        // Short circuit. An empty options hash does not error.
+        //
+        // Example:
+        //
+        // ```console
+        // [2.6.3]> Time.at(0, {})
+        // => 1970-01-01 01:00:00 +0100
+        // ```
+        if hash.is_empty() {
+            return Ok(None);
+        }
+
+        // All other keys are rejected.
+        //
+        // Example:
+        // ```console
+        // [2.6.3]> Time.at(0, i: 0)
+        // ArgumentError (unknown keyword: i)
+        // ```
+        for (mut key, _) in &hash {
+            let k = unsafe { Symbol::unbox_from_value(&mut key, self)? }.bytes(self);
+            if k != b"in" {
+                let mut message = b"unknown keyword: ".to_vec();
+                message.extend_from_slice(k);
+                Err(ArgumentError::from(message))?;
+            }
+        }
+
+        // Based on the above logic, the only option in the hash is `in`.
+        // >0 keys, and all other keys are rejected).
+        let mut in_value = hash.get(0).expect("Only the `in` parameter should be available").1;
+
+        match in_value.ruby_type() {
+            Ruby::String => {
```

Comment:
This doesn't look quite right.

```
[3.1.2] > class B; def to_str; 'B'; end; end
=> :to_str
[3.1.2] > Time.at(12, in: B.new)
=> 1970-01-01 02:00:12 +0200
[3.1.2] > class C; def to_str; raise 'crash'; end; end
=> :to_str
[3.1.2] > Time.at(12, in: C.new)
(irb):10:in `to_str': crash (RuntimeError)
        from <internal:timev>:274:in `at'
        from (irb):11:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > class D; def to_int; raise 'dint'; end; def to_str; raise 'dstr'; end; end
=> :to_str
[3.1.2] > Time.at(12, in: D.new)
(irb):12:in `to_str': dstr (RuntimeError)
        from <internal:timev>:274:in `at'
        from (irb):13:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

It looks like MRI prefers to implicitly convert through `#to_str` if it exists and otherwise will convert through `#to_int`. We can check for this by doing a `Value::respond_to` check.

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/2093